### PR TITLE
feat: bucket migrate check when load from db

### PR DESCRIPTION
**Pull request by jingjunLi**
### Description

refine bucket migrate

### Rationale

NA

### Example

NA

### Changes

Notable changes: 
* Add a cache for bucket info in BucketMigrateScheduler to avoid repeated chain queries.
* Store the DestGlobalVirtualGroupID field in MigrateGVGUnitMeta for restoring the plan during loading.
* Adapt ListMigrateBucketEvents for scenarios without completeEvents.
* Enhance loading from the database by verifying meta information on-chain during the check.

### CODE REVIEW

File: modular/manager/bucket_migrate_scheduler.go [hunk 1]
```diff
@@ -7,6 +7,8 @@ import (
 	"sync"
 	"time"
 
+	lru "github.com/hashicorp/golang-lru"
+
 	"cosmossdk.io/math"
```
Comment (sysvm) [inline modular/manager/bucket_migrate_scheduler.go:12]: sort imported pkg

File: modular/manager/bucket_migrate_scheduler.go [hunk 2]
```diff
@@ -311,17 +330,58 @@ func (s *BucketMigrateScheduler) Start() error {
 // Before processing MigrateBucketEvents, first check if the status of the bucket on the chain meets the expectations. If it meets the expectations, proceed with the execution; otherwise, skip this MigrateBucketEvent event.
 func (s *BucketMigrateScheduler) checkBucketFromChain(bucketId uint64, expectedStatus storagetypes.BucketStatus) (expected bool, err error) {
 	// check the chain's bucket is migrating
-	bucketInfo, err := s.manager.baseApp.Consensus().QueryBucketInfoById(context.Background(), bucketId)
-	if err != nil {
-		return false, err
+	key := cacheKey(bucketId)
+	var (
+		bucketInfo *storagetypes.BucketInfo
+	)
```
Comment (sysvm) [inline modular/manager/bucket_migrate_scheduler.go:336]: redundant parentheses

File: modular/manager/bucket_migrate_scheduler.go [hunk 3]
```diff
@@ -311,17 +330,58 @@ func (s *BucketMigrateScheduler) Start() error {
 // Before processing MigrateBucketEvents, first check if the status of the bucket on the chain meets the expectations. If it meets the expectations, proceed with the execution; otherwise, skip this MigrateBucketEvent event.
 func (s *BucketMigrateScheduler) checkBucketFromChain(bucketId uint64, expectedStatus storagetypes.BucketStatus) (expected bool, err error) {
```
Comment (sysvm) [inline modular/manager/bucket_migrate_scheduler.go:331]: fix `bucketId` as `bucketID`

File: modular/manager/bucket_migrate_scheduler.go [hunk 4]
```diff
@@ -311,17 +330,58 @@ func (s *BucketMigrateScheduler) Start() error {
...
 	}
+
 	if bucketInfo.BucketStatus != expectedStatus {
 		log.Debugw("the bucket status is not same, the event will skip", "bucketInfo", bucketInfo, "expectedStatus", expectedStatus)
 		return false, nil
 	}
 	return true, nil
 }
 
+func (s *BucketMigrateScheduler) doneMigrateBucket(bucketId uint64) error {
```
Comment (sysvm) [inline modular/manager/bucket_migrate_scheduler.go:359]: bucketId ditto

File: modular/manager/bucket_migrate_scheduler.go [hunk 5]
```diff
@@ -87,15 +94,16 @@ func (plan *BucketMigrateExecutePlan) storeToDB() error {
...
-			RedundancyIndex:      -1,
-			BucketID:             migrateGVGUnit.bucketID,
-			SrcSPID:              migrateGVGUnit.srcSP.GetId(),
-			DestSPID:             migrateGVGUnit.destSP.GetId(),
-			LastMigratedObjectID: migrateGVGUnit.lastMigratedObjectID,
-			MigrateStatus:        int(migrateGVGUnit.migrateStatus),
+			MigrateGVGKey:            migrateGVGUnit.Key(),
+			GlobalVirtualGroupID:     migrateGVGUnit.srcGVG.GetId(),
+			VirtualGroupFamilyID:     migrateGVGUnit.srcGVG.GetFamilyId(),
+			RedundancyIndex:          -1,
```
Comment (will-2012) [inline modular/manager/bucket_migrate_scheduler.go:98]: Refine  `-1` to `piecestore.PrimarySPRedundancyIndex`.

File: core/spdb/entity.go [hunk 6]
```diff
@@ -99,16 +99,17 @@ type OffChainAuthKey struct {
...
+	SwapOutKey               string
+	GlobalVirtualGroupID     uint32 // is used by sp exit/bucket migrate
+	VirtualGroupFamilyID     uint32 // is used by sp exit
+	RedundancyIndex          int32  // is used by sp exit
+	BucketID                 uint64 // is used by bucket migrate
+	SrcSPID                  uint32
+	DestSPID                 uint32
+	LastMigratedObjectID     uint64
+	MigrateStatus            int    // scheduler assign unit status.
+	DestGlobalVirtualGroupID uint32 // is used by bucket migrate
```
Comment (will-2012) [inline core/spdb/entity.go:112]: Move to the back of `GlobalVirtualGroupID`.

File: modular/manager/bucket_migrate_scheduler.go [hunk 7]
```diff
@@ -297,6 +311,11 @@ func (s *BucketMigrateScheduler) Init(m *ManageModular) error {
 	}
 	s.executePlanIDMap = make(map[uint64]*BucketMigrateExecutePlan)
 
+	s.bucketCache, err = lru.New(bucketCacheSize)
```
Comment (will-2012) [inline modular/manager/bucket_migrate_scheduler.go:312]: What expires in the cache in the LRU, if it has not expired, is there any problem?

File: modular/manager/bucket_migrate_scheduler.go [hunk 8]
```diff
@@ -538,30 +575,63 @@ func (s *BucketMigrateScheduler) replaceExitingSP(secondarySPIDs []uint32) ([]ui
...
 		}
-		destSP, queryErr := s.manager.virtualGroupManager.QuerySPByID(migrateGVG.DestSPID)
-		if queryErr != nil {
-			log.Errorw("failed to query sp", "error", queryErr, "sp_info", destSP)
-			return nil, queryErr
+		executePlan, err = s.generateBucketMigrateGVGExecuteUnit(primarySPGVGList, event, executePlan, srcSP, destSP)
+		if err != nil {
+			return nil, err
 		}
+	} else {
```
Comment (will-2012) [inline modular/manager/bucket_migrate_scheduler.go:606]: What if the quantity is the same but the content is different?

File: modular/manager/bucket_migrate_scheduler.go [hunk 9]
```diff
@@ -309,19 +326,58 @@ func (s *BucketMigrateScheduler) Start() error {
...
 	// check the chain's bucket is migrating
-	bucketInfo, err := s.manager.baseApp.Consensus().QueryBucketInfoById(context.Background(), bucketId)
-	if err != nil {
-		return false, err
+	key := cacheKey(bucketID)
+	var bucketInfo *storagetypes.BucketInfo
+	elem, has := s.bucketCache.Get(key)
+	if has {
+		value, ok := elem.(storagetypes.BucketInfo)
+		if !ok {
```
Comment (will-2012) [inline modular/manager/bucket_migrate_scheduler.go:336]: Under what conditions does this branch execute? Add more log?

Comment (will-2012) [reply modular/manager/bucket_migrate_scheduler.go:336]: Maybe need query chain to get the latest bucket info.

File: core/spdb/entity.go (hunk 6 above)
Comment (will-2012) [reply core/spdb/entity.go:112]: Need to add db implementation.

File: core/vgmgr/virtual_group_manager.go [hunk 10]
```diff
@@ -41,6 +42,12 @@ type GVGPickFilter interface {
 	CheckGVG(gvgMeta *GlobalVirtualGroupMeta) bool
 }
 
+// GVGMetaChecker is used to check sp pick condition.
+type GVGMetaChecker interface {
+	// CheckGVG returns true when match pick request condition.
+	CheckGVG(migrateGVGUnitMeta []*spdb.MigrateGVGUnitMeta) bool
```
Comment (will-2012) [inline core/vgmgr/virtual_group_manager.go:48]: The interfaces defined here are all used in base/gfspvgmr. The interface of GVGMetaChecker should not be placed here.

File: modular/manager/bucket_migrate_scheduler.go [hunk 11]
```diff
@@ -309,19 +363,65 @@ func (s *BucketMigrateScheduler) Start() error {
...
+			log.Debugw("failed to get bucketinfo from bucket cache", "key", key)
+			s.bucketCache.Delete(key)
+			QueryBucketInfoFromChainFunc()
+		}
+		bucketInfo = &value
+	} else {
+		QueryBucketInfoFromChainFunc()
+	}
+
 	if bucketInfo.BucketStatus != expectedStatus {
```
Comment (will-2012) [inline modular/manager/bucket_migrate_scheduler.go:392]: bucketInfo is nil when query failed.

File: modular/manager/bucket_migrate_scheduler.go [hunk 12]
```diff
@@ -61,9 +65,38 @@ func (f *PickDestGVGFilter) CheckGVG(gvgMeta *vgmgr.GlobalVirtualGroupMeta) bool
 	return false
 }
 
+// GVGMetaCheck verifies whether expectedSPGVGList completely matches with the migrateGVGUnitMeta loaded from the database.
+func GVGMetaCheck(expectedSPGVGList []*virtualgrouptypes.GlobalVirtualGroup, migrateGVGUnitMeta []*spdb.MigrateGVGUnitMeta) bool {
```
Comment (will-2012) [inline modular/manager/bucket_migrate_scheduler.go:69]: `CheckGVGMetaConsistent` may be a good function name.

File: modular/manager/bucket_migrate_scheduler.go [hunk 13]
```diff
@@ -61,9 +65,38 @@ func (f *PickDestGVGFilter) CheckGVG(gvgMeta *vgmgr.GlobalVirtualGroupMeta) bool
 	return false
 }
 
+// CheckGVGMetaConsistent verifies whether expectedSPGVGList completely matches with the migrateGVGUnitMeta loaded from the database.
+func CheckGVGMetaConsistent(expectedSPGVGList []*virtualgrouptypes.GlobalVirtualGroup, migrateGVGUnitMeta []*spdb.MigrateGVGUnitMeta) bool {
```
Comment (will-2012) [inline modular/manager/bucket_migrate_scheduler.go:69]: expectedSPGVGList-->chainMetaList
migrateGVGUnitMeta-->dbMetaList

**Review by sysvm — APPROVED** [review]
LGTM

File: modular/manager/bucket_migrate_scheduler.go [hunk 14]
```diff
@@ -61,9 +65,38 @@ func (f *PickDestGVGFilter) CheckGVG(gvgMeta *vgmgr.GlobalVirtualGroupMeta) bool
...
+		for _, migrateGVG := range migrateGVGUnitMeta {
+			srcGvg, ok := chainGvgMaps[migrateGVG.GlobalVirtualGroupID]
+			if !ok {
+				return false
+			}
+			existMaps[srcGvg.GetId()] = true
+		}
+
+		for _, exist := range existMaps {
+			if !exist {
```
Comment (will-2012) [inline modular/manager/bucket_migrate_scheduler.go:86]: Will not go to this condition.

File: modular/manager/bucket_migrate_scheduler.go [hunk 15]
```diff
@@ -297,6 +337,11 @@ func (s *BucketMigrateScheduler) Init(m *ManageModular) error {
 	}
 	s.executePlanIDMap = make(map[uint64]*BucketMigrateExecutePlan)
 
+	s.bucketCache = NewBucketCache(bucketCacheSize, 30*time.Minute)
```
Comment (will-2012) [inline modular/manager/bucket_migrate_scheduler.go:340]: Put `30` to file const scope.

**Review by will-2012 — APPROVED** [review]

File: modular/manager/bucket_migrate_scheduler.go [hunk 16]
```diff
@@ -535,33 +614,66 @@ func (s *BucketMigrateScheduler) replaceExitingSP(secondarySPIDs []uint32) ([]ui
...
+	// 1) check chain's bucket status
+	expected, err := s.checkBucketFromChain(bucketID, storagetypes.BUCKET_STATUS_MIGRATING)
+	if err != nil {
+		return nil, err
+	}
+	if !expected {
+		return nil, nil
+	}
+
+	chainGvgMaps := make(map[uint32]*virtualgrouptypes.GlobalVirtualGroup)
```
Comment (Fynnss) [inline modular/manager/bucket_migrate_scheduler.go:631]: If the map size is known, you can directly make the memory of the corresponding size to improve performance

```
chainGvgMaps := make(map[uint32]*virtualgrouptypes.GlobalVirtualGroup, len(primarySPGVGList))
```

**Pull request merged**